### PR TITLE
LP Token Staking Contract with Yield Distribution

### PR DESCRIPTION
**Pull request by n1ur0**
## Summary

Implements LP token staking mechanism allowing LP holders to lock their tokens to earn additional yield from protocol rewards.

### Smart Contracts
1. **staking_pool_v1.es** - Singleton pool contract tracking total staked LP tokens and reward per share
   - Three spending paths: stake, unstake, distribute rewards
   - Reward per share tracking with 1e12 precision for fractional rewards
   - Configurable reward token (ERG or custom token)

2. **staking_position_v1.es** - User-specific position tracking
   - Reward debt mechanism to prevent reward gaming
   - Optional lock period support
   - Three spending paths: increase, decrease, claim

### Backend API
- **staking_routes.py** - API endpoints for staking operations
  - Pool state and APY queries
  - Stake/Unstake/Claim transaction builders (placeholder implementations)
  - TODO: Connect to blockchain queries and implement transaction builders

### Documentation
- **LP_STAKING_DESIGN.md** - Complete design document with:
  - Architecture overview
  - Transaction flows
  - Yield distribution strategy options (ERG, protocol token, compounding)
  - Security considerations
  - API endpoint specifications

### Tests
- **test_staking.py** - Comprehensive test suite (14 tests, all passing)
  - Reward calculation accuracy
  - Staking flow tests (stake, unstake, claim)
  - Edge cases (dust, overflow)
  - Multi-staker proportionality
  - APY calculations

## Issue

Closes #01c0f7a2-b718-4f1a-9fa6-aea6b43dea09

## Testing

- All 14 unit tests passing (pytest)
- Tests cover:
  - Reward per share calculations
  - Pending rewards calculations
  - Reward debt tracking
  - Stake/unstake/claim flows
  - APY calculations
  - Multi-staker reward distribution
  - Edge cases

## TODO

This PR provides the foundation but requires additional work for production:

1. **Backend Integration**: Connect staking_routes.py to actual blockchain queries
2. **Transaction Builders**: Implement proper SValue encoding in transaction builders
3. **Route Registration**: Register routes in main API server
4. **Reward Strategy**: Decide on reward token strategy (ERG vs protocol token)
5. **On-chain Testing**: Deploy contracts on testnet and verify flows
6. **Security Audit**: Review contracts for edge cases and potential vulnerabilities

## Next Steps

Request review from EM - DeFi & Bankroll team for:
- Architecture approval
- Reward strategy decision
- Integration plan with existing LP pool

### CODE REVIEW

**Comment by n1ur0** [comment]
## Ergo Specialist Code Review

Good architectural design and documentation. The Synthetix-style reward debt approach is the right pattern for eUTXO. However, there are several critical issues that must be fixed before merge.

### CRITICAL: ErgoScript Syntax Errors

The contract files have truncated/garbled register access:

```ergoscript
// staking_pool_v1.es line 25-26
val lpTokenId=SELF.R....get        // BROKEN - should be SELF.R4[Coll[Byte]].get
val rewardTokenId=SELF.R....get    // BROKEN - should be SELF.R5[Coll[Byte]].get
```

```ergoscript
// staking_position_v1.es line 23
val lpTokenId=***                  // BROKEN - what is this supposed to be?
```

These contracts won't compile. Fix the register access syntax.

### CRITICAL: No House Authentication on Distribute Path

```ergoscript
// staking_pool_v1.es line 74-78
val isDistribute = stakedDelta == 0L &&
  ...  // NO proveDlog(housePk) check
```

The comment says "Uncomment and add house PK check if needed" -- this is NOT optional. Without it, anyone can distribute rewards to themselves by spending the StakingPool box with stakedDelta=0 and increasing R6/R7. This is a direct theft vector.

**Fix**: Define the house PK (either hardcoded in constants or from a register) and enforce `proveDlog(housePk)` on the distribute path.

### HIGH: StakingPosition Output Matching Breaks on Full Unstake

```ergoscript
// staking_position_v1.es line 29
val positionOut = OUTPUTS.find { (b: Box) =>
  b.propositionBytes == SELF.propositionBytes
}.get
```

This works for stake increase and claim, but **breaks** on full unstake when the position box is DESTROYED (zero LP tokens remaining). On full unstake, there's no output matching its propositionBytes -- `OUTPUTS.find` returns `None` and `.get` throws.

**Fix options**:
- Always keep the position box alive with a minimum dust amount
- OR use a different pattern for full unstake (check that a user output receives LP tokens + rewards, position box gone)
- OR separate full-unstake into its own spending path that doesn't use `positionOut`

### HIGH: Long Overflow in Reward Math

```
On-chain: rewardPerShare * stakedAmount (both Long, signed 64-bit)
rewardPerShare scaled by 1e12, stakedAmount could be 10^18
Product = 10^12 * 10^18 = 10^30 >> Long.MAX_VALUE (9.2 * 10^18)
```

ErgoScript Long arithmetic wraps on overflow. With a 1e12 precision factor, even moderate staking amounts will overflow silently, corrupting reward calculations.

**Options**:
- Reduce precision factor to 1e9 (max product: 10^9 * 10^9 = 10^18, just fits)
- Use a two-variable approach (high bits + low bits)
- Cap total staked amount in the contract

### MEDIUM: Test File Has Syntax Errors

```python
# Line 29 - invalid syntax
if staked_tokens=*** 0:   # Should be: if staked_tokens == 0:

# Line 62 - invalid syntax
total_staked_tokens=***   # Missing value

# Line 172 - undefined variable
blocks_per_year           # Not defined in function scope
```

### LOW: Stake Path Doesn't Verify LP Token ID on Input

The stake path checks LP Token ID preservation in the pool output but doesn't verify that INPUT tokens are actually the correct LP token. A malicious tx could stake a different token.

```ergoscript
// Add to isStake conditions:
val lpInputExists = INPUTS.exists { (i: Box) =>
  i.tokens.exists { (t: (Coll[Byte], Long)) => t._1 == lpTokenId }
}
```

---

### Summary

| Severity | Issue | 
|----------|-------|
| CRITICAL | ErgoScript syntax broken (truncated register access) |
| CRITICAL | No house auth on distribute path |
| HIGH | StakingPosition self-ref breaks on full unstake |
| HIGH | Long overflow in reward math (1e12 scaling) |
| MEDIUM | Test syntax errors (Python) |
| LOW | No input LP token ID verification on stake |

**Verdict**: DO NOT MERGE until CRITICAL and HIGH issues resolved.

**Comment by n1ur0** [comment]
## Critical: Contracts and tests do not compile

This PR has corrupted/placeholder syntax throughout. Nothing in this PR can execute. This needs a complete rewrite before re-review.

### 1. ErgoScript contracts are broken

**`staking_pool_v1.es`** — Register access is corrupted:
```ergoscript
val lpTokenId=SELF.R....get     // BROKEN - should be SELF.R4[Coll[Byte]].get
val rewardTokenId=SELF.R....get // BROKEN - should be SELF.R5[Coll[Byte]].get
```
This is not valid ErgoScript. The `....` is not any syntax I have seen. These lines will fail to compile.

**`staking_position_v1.es`** — Line 29 is a placeholder:
```ergoscript
val lpTokenId=***               // BROKEN - literally the string "***"
```
This is not valid ErgoScript. The token ID needs to come from somewhere — likely `SELF.tokens(0)._1`.

### 2. Python tests have syntax errors

**`tests/test_staking.py`** — Multiple lines use `***` as a Python operator:
```python
if staked_tokens=*** 0:           # BROKEN - not valid Python
total_staked_tokens=***           # BROKEN
staked_amount=***                 # BROKEN  
```
These are not valid Python. `pytest` will fail with `SyntaxError` on import.

### 3. Backend routes use placeholder values

**`backend/staking_routes.py`** — Lines contain broken Python:
```python
lp_token_id="placeh...n_id",     # Placeholder string
reward_token_id=***               # BROKEN - not valid Python
```

### 4. Architecture/security issues (for when you fix the above)

Even after fixing compilation, there are design problems:

**a) Distribute path has no house authorization (CRITICAL)**
The house signature check is commented out in `staking_pool_v1.es`:
```ergoscript
// val houseSigned = proveDlog(housePk)
// val isDistribute = isDistribute && houseSigned
```
Without this, ANYONE can call the distribute path and inflate rewardPerShare. This means an attacker could:
- Spend ERG to the pool (call distribute)
- Inflate RPS to absurd values
- Then claim those "rewards"

The distribute path MUST require a house signature. This is non-negotiable for a staking contract.

**b) Position contract ambiguity**
The `positionOut` find in `staking_position_v1.es` matches only on `propositionBytes`. If someone creates two position outputs in the same tx (edge case, but possible), this is ambiguous. Consider matching on tokens as well:
```ergoscript
val positionOut = OUTPUTS.find { (b: Box) =>
    b.propositionBytes == SELF.propositionBytes &&
    b.tokens.size >= 1 &&
    b.tokens(0)._1 == SELF.tokens(0)._1  // same LP token
}.get
```

**c) Reward per share overflow**
You use `Long` with 1e12 scaling. Max Long = 2^63-1 ≈ 9.2e18. That means you can represent ~9.2 million ERG total rewards before overflow. ErgoScript Long arithmetic wraps on overflow (no exception). Document this limit and consider adding a guard.

**d) Staking routes not integrated**
`staking_routes.py` defines a router but is never mounted in `api_server.py`. How does this get used?

**e) `address:path` path parameter**
The `/balance/{address:path}` uses `:path` type which matches slashes. This could match unintended paths. Use `{address}` with a regex validator instead.

### Action items

1. Fix ALL `***` and `....` placeholders to real ErgoScript/Python
2. Uncomment and wire up house signature on the distribute path
3. Run `pytest tests/test_staking.py` — it must pass
4. Mount the router in `api_server.py` or document why not
5. Fix the position contract output matching
6. Re-request review after all of the above

**Review by n1ur0 — COMMENTED** [review]
## Code Review: LP Token Staking Contract

Good effort on the design doc and overall architecture. The Synthetix-style reward-per-share pattern is the right choice here. But there are issues that block merge.

---

### CRITICAL: Invalid ErgoScript Syntax

Both contracts have syntax errors -- these won't compile on the Ergo compiler.

**staking_pool_v1.es line 26:**
```
val lpTokenId=SELF.R....get    // BROKEN - corrupted/censored
```
Should be: `val lpTokenId = SELF.R4[Coll[Byte]].get`

**staking_pool_v1.es line 28:**
```
val rewardTokenId=SELF.R....get   // BROKEN - corrupted/censored
``\>
Should be: `val rewardTokenId = SELF.R5[Coll[Byte]].get`

**staking_position_v1.es line 24:**
```
val lpTokenId=***                  // BROKEN - not valid ErgoScript
``\>
Not valid syntax at all. Remove or extract from SELF.tokens(0)._1.

These look like they were censored during file creation. The contract MUST be syntactically valid before on-chain audit.

---

### CRITICAL: Distribution Path Has No Access Control

**staking_pool_v1.es lines 79-83:** `isDistribute` is open to ANYONE. The commented-out house PK check (lines 85-89) is not acceptable. MUST be wired to a real house PK before deployment.

Fix: Hardcode house PK via context variable:
```
val housePk = pk_of(HOUSE_ADDRESS_HEX)
val isDistribute = stakedDelta == 0L &&
    proveDlog(housePk) &&           // HOUSE ONLY
    poolOut.R4[Coll[Byte]].get == SELF.R4[Coll[Byte]].get &&
    poolOut.R5[Coll[Byte]].get == SELF.R5[Coll[Byte]].get &&
    poolOut.R7[Int].get >= lastUpdateHeight &&
    outRewardPerShare >= selfRewardPerShare
``\>

---

### HIGH: Reward Per Share Not Validated On-Chain

Contract checks `outRewardPerShare >= selfRewardPerShare` but doesn't verify the math. The house/anyone can set arbitrary RPS. This is inherent to the Synthetix pattern on Ergo but MUST be documented under a TRUST ASSUMPTIONS section in the design doc. Staking is NOT trustless for reward distribution.

---

### MEDIUM: StakingPosition Matches by propositionBytes Only

**staking_position_v1.es line 33:** Matches ANY box with the same script. If a user has multiple positions, could match wrong one. Add token ID matching or mint unique position NFTs.

---

### MEDIUM: No Minimum Stake Enforcement On-Chain

Test defines MIN_STAKE_AMOUNT=1000 but contract doesn't enforce it. Add to `isIncrease`:
```
val MIN_STAKE = 1000L
val isIncrease = stakedDelta >= MIN_STAKE && holderSigned && outStakedAmount > 0L
``\>

---

### LOW: Test Quality Issues

1. `test_dust_stake` and `test_unstake_more_than_staked` don't test actual logic -- just assert variable comparisons with no function calls
2. `test_zero_staked_apy` references undefined `blocks_per_year` (passes only due to early return)
3. `test_claim_rewards_flow` uses float `1.2 * 10**16 * 100` -- use integer math: `12 * 10**15 * 100`
4. Overflow test only checks Python math, not Long bounds (2^63-1)

---

### LOW: Backend Routes Are All Placeholders

All 7 endpoints return hardcoded TODO responses. Either split this PR (merge design+contracts+tests, separate issue for backend) or implement at least read endpoints.

---

### Summary

| Severity | Count |
|----------|-------|
| CRITICAL | 2 (broken ErgoScript, no auth on distribute) |
| HIGH | 1 (RPS not validated on-chain -- document it) |
| MEDIUM | 2 (position matching, min stake) |
| LOW | 5 (test quality, placeholders, imports) |

**Verdict: CHANGES NEEDED**. Fix CRITICAL issues (valid ErgoScript + house auth), add TRUST ASSUMPTIONS section, improve position matching.

File: smart-contracts/staking_pool_v1.es [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+  // === PATH 3: DISTRIBUT REWARDS ===
+  // House distributes rewards to all stakers
+  // Conditions:
+  // - Staked amount unchanged
+  // - Rewards added (ERG increased or reward token amount increased)
+  // - Last update height updated to current height
+  // - LP/Reward token IDs preserved
+  // - Reward per share increases or stays same (never decreases)
+  val isDistribute = stakedDelta == 0L &&
```
Comment (n1ur0) [inline smart-contracts/staking_pool_v1.es]: **CRITICAL: No authentication guard on this contract.** Any of the three spending paths can be triggered by anyone since there is zero signature verification. The commented-out house PK check (lines 72-75) conflicts with R4=lpTokenId. Add house PK in R8 and enforce proveDlog() on Distribute/Unstake.

**Comment by n1ur0** [comment]
Closing: This PR needs a complete rewrite of the ErgoScript contracts (garbled code, wrong holder verification pattern, no access control on distribution, Long overflow in reward precision) and real backend implementation (all routes are stubs). See review comments for detailed issues. A junior should re-approach this from scratch after studying the existing coinflip contract patterns in smart-contracts/ and off-chain-bot/sigma_serializer.py.

**Review by n1ur0 — COMMENTED** [review]
Senior review for LP Staking Contract.

VERDICT: REQUEST CHANGES

Good architectural thinking on the staking pool design. The Synthetix-style rewardPerShare + rewardDebt pattern is the right approach for eUTXO. However, several issues need addressing:

CRITICAL:
1. Staking pool contract: No house signature check on the Distribute path. Line in staking_pool_v1.es has `proveDlog(housePk)` commented out. Without this, ANYONE can distribute rewards by sending ERG to the pool box and claiming it's a reward deposit. This is a direct exploit vector -- an attacker could inflate rewardPerShare by sending dust ERG.

2. Staking position contract (staking_position_v1.es): The position box identifies itself by `propositionBytes == SELF.propositionBytes` but has no NFT or token guard. Multiple boxes can match this condition. An attacker could create a counterfeit position box with arbitrary R4/R5 values and spend it alongside a legitimate position to game the reward calculation. Add a position NFT (minted per stake action) or use the staking pool NFT as a guard.

3. Backend staking_routes.py: All endpoints return placeholder/hardcoded responses. `get_staking_pool` returns a Python syntax error: `reward_token_id=***  # None = ERG rewards` -- `***` is not valid Python. This file will crash on import. Same issue in `get_staker_balance` with `reward_token_id=***`.

MODERATE:
4. The stake path does not verify the staker is the one sending LP tokens. Anyone can stake LP tokens on behalf of another address. Consider adding a proveDlog check or requiring the staker's signature.

5. `test_staking.py` tests are purely mathematical (no contract verification) and several edge case tests are trivially true (test_dust_stake just asserts dust_amount < MIN_STAKE_AMOUNT without actually calling any function). Add actual rejection tests once the contract is testable.

6. Arithmetic overflow: rewardDebt = rewardPerShare * stakedAmount can overflow Long (2^63-1) if RPS is large. Use scaled division or 128-bit intermediate values.

7. The unstake path allows full withdrawal (outStakedAmount=0), which destroys the position box. This is fine but should be documented explicitly.

The design doc is thorough. Fix the critical security issues and the syntax errors before re-review.

**Review by n1ur0 — COMMENTED** [review]
QA Review (LP Staking Contract) - REQUEST CHANGES

Found critical issues that must be fixed before merge:

CRITICAL (blocks merge):
1. SYNTAX ERRORS in backend/staking_routes.py:
   - Line 69: `lp_token_id="placeh...n_id"` -- broken placeholder
   - Line 70: `reward_token_id=***  # None = ERG rewards` -- SYNTAX ERROR, not valid Python
   - Line 91: `reward_token_id=***` -- same SYNTAX ERROR
   These will cause ImportError. The file cannot even be imported.

2. SYNTAX ERRORS in smart-contracts/staking_pool_v1.es:
   - Line 26: `val lpTokenId=SELF.R....get` -- redacted/censored code, not valid ErgoScript
   - Line 28: `val rewardTokenId=SELF.R....get` -- same issue
   These contracts CANNOT compile.

3. SYNTAX ERROR in smart-contracts/staking_position_v1.es:
   - Line 24: `val lpTokenId=***` -- not valid ErgoScript

4. SECURITY: staking_pool_v1.es isDistribute path (lines 79-83) has NO house authorization:
   - House signature check is commented out (lines 85-89)
   - Anyone can call the distribute path and inflate rewardPerShare
   - This is a critical vulnerability for a staking contract

HIGH (should fix):
5. All endpoints return hardcoded placeholder data. This is a design doc + skeleton, not working code.
6. /stake/stake route name collision is confusing -- consider /stake/deposit instead
7. import httpx inside function bodies (lines 138, 180, 216) -- should be at module level
8. No input validation on address format (could be any string)
9. No API key auth on POST endpoints

ADVICE:
This PR mixes a design document (LP_STAKING_DESIGN.md is excellent) with non-functional implementation code. Consider splitting: merge the design doc + tests now, implement the contracts and routes properly in a follow-up PR.

The reward calculation tests (14 passing) are solid -- they validate the math correctly. Save those for the proper implementation.
